state/machine_test.go

```diff
@@ -2419,3 +2420,158 @@ func (s *MachineSuite) TestMachineAddDifferentAction(c *gc.C) {
+		u, err := s.State.Unit(p)
+		c.Assert(err, jc.ErrorIsNil)
+		c.Assert(u.Series(), gc.Equals, series)
+		subs := u.SubordinateNames()
+		for _, sn := range subs {
+			u, err := s.State.Unit(sn)
+			c.Assert(err, jc.ErrorIsNil)
+			c.Assert(u.Series(), gc.Equals, series)
+		}
+	}
+}
+
+func (s *MachineSuite) TestUpdateMachineSeries(c *gc.C) {
+	mach := s.setupTestUpdateMachineSeries(c)
+	err := mach.UpdateMachineSeries("trusty", false)
+	c.Assert(err, jc.ErrorIsNil)
+	s.assertMachineAndUnitSeriesChanged(c, mach, "trusty")
+}
+
+func (s *MachineSuite) TestUpdateMachineSeriesFail(c *gc.C) {
+	mach := s.setupTestUpdateMachineSeries(c)
+	err := mach.UpdateMachineSeries("xenial", false)
+	c.Assert(err, jc.Satisfies, state.IsIncompatibleSeriesError)
+	s.assertMachineAndUnitSeriesChanged(c, mach, "precise")
+}
+
+func (s *MachineSuite) TestUpdateMachineSeriesForce(c *gc.C) {
+	mach := s.setupTestUpdateMachineSeries(c)
+	err := mach.UpdateMachineSeries("xenial", true)
+	c.Assert(err, jc.ErrorIsNil)
+	s.assertMachineAndUnitSeriesChanged(c, mach, "xenial")
+}
+
+func (s *MachineSuite) TestUpdateMachineSeriesSameSeriesToStart(c *gc.C) {
+	mach := s.setupTestUpdateMachineSeries(c)
+	err := mach.UpdateMachineSeries("precise", false)
+	c.Assert(err, jc.ErrorIsNil)
+	s.assertMachineAndUnitSeriesChanged(c, mach, "precise")
+}
+
+func (s *MachineSuite) TestUpdateMachineSeriesSameSeriesAfterStart(c *gc.C) {
+	mach := s.setupTestUpdateMachineSeries(c)
+
+	defer state.SetTestHooks(c, s.State,
+		jujutxn.TestHook{
+			Before: func() {
+				ops := []txn.Op{{
+					C:      state.MachinesC,
+					Id:     state.DocID(s.State, mach.Id()),
+					Update: bson.D{{"$set", bson.D{{"series", "trusty"}}}},
+				}}
+				err := state.RunTransaction(s.State, ops)
+				c.Assert(err, jc.ErrorIsNil)
+			},
+			After: func() {
+				err := mach.Refresh()
+				c.Assert(err, jc.ErrorIsNil)
+				c.Assert(mach.Series(), gc.Equals, "trusty")
+			},
+		},
+	).Check()
+
+	err := mach.UpdateMachineSeries("trusty", false)
+	c.Assert(err, jc.ErrorIsNil)
+	err = mach.Refresh()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(mach.Series(), gc.Equals, "trusty")
+}
+
+func (s *MachineSuite) TestUpdateMachineSeriesCharmURLChangedSeriesFail(c *gc.C) {
+	mach := s.setupTestUpdateMachineSeries(c)
+
+	defer state.SetTestHooks(c, s.State,
+		jujutxn.TestHook{
+			Before: func() {
+				v2 := state.AddTestingCharmMultiSeries(c, s.State, "multi-seriesv2")
+				cfg := state.SetCharmConfig{Charm: v2}
+				app, err := s.State.Application("multi-series")
+				c.Assert(err, jc.ErrorIsNil)
+				err = app.SetCharm(cfg)
+				c.Assert(err, jc.ErrorIsNil)
+			},
+		},
+	).Check()
+
+	// Trusty is listed in only version 1 of the charm.
+	err := mach.UpdateMachineSeries("trusty", false)
+	c.Assert(err, gc.ErrorMatches, "cannot update series for \"2\" to trusty: series \"trusty\" not supported by charm, supported series are: precise,xenial")
+}
+
+func (s *MachineSuite) TestUpdateMachineSeriesPrincipalsListChange(c *gc.C) {
```

Comment:
Let's add a test for subordinate changes also. I think my preference would be for a new principal/subordinate added which introduces an incompatible series and the test asserts that error.

Reply:
added.